crates/core_arch/src/wasm32/simd128.rs

```diff
@@ -143,350 +149,738 @@ extern "C" {
+pub const unsafe fn i64x2_const<const A0: i64, const A1: i64>() -> v128 {
+    transmute(i64x2(A0, A1))
 }
 
-/// Creates a vector with identical lanes.
+#[cfg(all(test, all_simd))]
+#[assert_instr(v128.const)]
+#[target_feature(enable = "simd128")]
+unsafe fn test_i64x2_const() -> v128 {
+    i64x2_const::<0, 2>()
+}
+
+/// Materializes a constant SIMD value from the immediate operands.
 ///
-/// Construct a vector with `x` replicated to all 8 lanes.
+/// This function generates a `v128.const` instruction as if the generated
+/// vector was interpreted as four 32-bit floats.
 #[inline]
-#[cfg_attr(test, assert_instr(i16x8.splat))]
-pub fn i16x8_splat(a: i16) -> v128 {
-    unsafe { transmute(i16x8::splat(a)) }
+#[target_feature(enable = "simd128")]
+pub const unsafe fn f32x4_const<const A0: f32, const A1: f32, const A2: f32, const A3: f32>() -> v128
+{
+    transmute(f32x4(A0, A1, A2, A3))
 }
 
-/// Extracts a lane from a 128-bit vector interpreted as 8 packed i16 numbers.
+#[cfg(all(test, all_simd))]
+#[assert_instr(v128.const)]
+#[target_feature(enable = "simd128")]
+unsafe fn test_f32x4_const() -> v128 {
+    f32x4_const::<0.0, 1.0, 0.0, 0.0>()
+}
+
+/// Materializes a constant SIMD value from the immediate operands.
 ///
-/// Extracts a the scalar value of lane specified in the immediate mode operand
-/// `imm` from `a`.
+/// This function generates a `v128.const` instruction as if the generated
+/// vector was interpreted as two 64-bit floats.
+#[inline]
+#[target_feature(enable = "simd128")]
+pub const unsafe fn f64x2_const<const A0: f64, const A1: f64>() -> v128 {
+    transmute(f64x2(A0, A1))
+}
+
+#[cfg(all(test, all_simd))]
+#[assert_instr(v128.const)]
+#[target_feature(enable = "simd128")]
+unsafe fn test_f64x2_const() -> v128 {
+    f64x2_const::<0.0, 1.0>()
+}
+
+/// Returns a new vector with lanes selected from the lanes of the two input
+/// vectors `$a` and `$b` specified in the 16 immediate operands.
+///
+/// The `$a` and `$b` expressions must have type `v128`, and this macro
+/// generates a wasm instruction that is encoded with 16 bytes providing the
+/// indices of the elements to return. The indices `i` in range [0, 15] select
+/// the `i`-th element of `a`. The indices in range [16, 31] select the `i -
+/// 16`-th element of `b`.
+///
+/// Note that this is a macro due to the codegen requirements of all of the
+/// index expressions `$i*` must be constant. A compiler error will be
+/// generated if any of the expressions are not constant.
+///
+/// All indexes `$i*` must have the type `u32`.
+#[inline]
+#[target_feature(enable = "simd128")]
+pub unsafe fn v8x16_shuffle<
+    const I0: u32,
+    const I1: u32,
+    const I2: u32,
+    const I3: u32,
+    const I4: u32,
+    const I5: u32,
+    const I6: u32,
+    const I7: u32,
+    const I8: u32,
+    const I9: u32,
+    const I10: u32,
+    const I11: u32,
+    const I12: u32,
+    const I13: u32,
+    const I14: u32,
+    const I15: u32,
```

Comment:
I feel that it would be better to make these indicies `usize` (for consistency with array indices) unless there is a good reason to make them `u32`.